### DISCUSSION TITLE
Solving the recursive RAVs problem

### DISCUSSION THREAD

**Discussion by coffeeDev98** [mliebelt/pgn-viewer#380] [Q&A] [answered]
Hey @mliebelt ,

Was curious on how you managed to resolve the Variations recurring at multiple levels... for example:

`1. e4 d5 
   ( 1... c5 2. e5 
     ( 2. b4 cxb4 
     ( 2... a5 3. bxa5 c4 
        ( 3... b5 4. axb6 
           ( 4. e5 )....
        )
      ) 
      )
    ) *`

From going through the code, I can see `pgn-reader` handles the major portion of JSON traversal & manipulations after PGN is parsed using `pgn-parser`...  but I wanted to know how you handled the possible situations where many variations are created one after the other...

Hope the above example explains my query...

**Comment by mliebelt** [comment, marked as answer]
Well, it is mostly an implementation done in the grammar, with some additions done later when reading it. So here are the steps:

1. Store your pgn in a file ( I named it `file.pgn` ).
2. Call then `pgn-parser file.pgn` and get as result the JSON it creates. (or use `pgn-parser -p file.pgn` to get it pretty-printed). You see that the structure is all there ... When you look at it, you will see that it is an array of moves. The file [`types.ts`](https://github.com/mliebelt/pgn-parser/blob/11a9dccb5bdb60417a9fce609b7246db132214b3/src/types.ts#L34) shows the type `PgnMove` with its structure. Part of it is (recursively) an array  of array of `PgnMove`s again (named `variations`). And that is the main story.
3. The `pgn-reader` now uses that call of functions: `loadOne --> readMoves --> eachMove --> eachMoveVariation` now does the cleanup: adding information missing after the parsing, _wiring_ moves to each other (so that each move knows its predecessor, ... Very complicated, I am not proud on it, but it works.

Long story short: the grammar does 95% of the stuff, and the JSON emitted should be sufficient to do whatever you want to do. So the answer to your question is: by using recursion, and having a recursive structure build by the parser.

I will attach the example files: `file.pgn`, `file.json` (result after parsing). Had to rename them, because GitHub does not allow those extensions.
[file.txt](https://github.com/mliebelt/PgnViewerJS/files/8911966/file.txt)
[file-result.txt](https://github.com/mliebelt/PgnViewerJS/files/8911967/file-result.txt)
.
